File: sdbot/signature.py
"""Webhook signature verification and signing helpers.

GitHub (baseline): ``X-Hub-Signature-256: sha256=<hex HMAC-SHA256 of the raw body>``.
GitCode: ``X-GitCode-Signature-256: sha256=<HMAC-SHA256 of the raw body>`` in "signature"
mode, or ``X-GitCode-Token: <password>`` in "password" mode. GitCode's docs do not say
whether the digest is hex or base64, so both encodings are accepted (each is still a
full MAC over the body with the shared secret). The legacy SHA-1 ``X-Hub-Signature`` is
ignored on purpose.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
from dataclasses import dataclass

GITHUB_SIGNATURE_HEADER = "X-Hub-Signature-256"
GITCODE_SIGNATURE_HEADER = "X-GitCode-Signature-256"
GITCODE_TOKEN_HEADER = "X-GitCode-Token"


@dataclass(frozen=True)
class Verification:
    ok: bool
    mode: str        # hmac-sha256 | token | unsigned | rejected
    reason: str = ""


def header(headers, name: str) -> str | None:
    """Case-insensitive header lookup that works for ``email.message.Message`` and plain dicts."""
    value = headers.get(name)
    if value is None:
        wanted = name.lower()
        for key, candidate in headers.items():
            if key.lower() == wanted:
                return candidate
    return value


def verify(provider: str, headers, body: bytes, secret: str | None) -> Verification:
    """Check the delivery against ``secret``. No secret configured means the delivery is
    accepted unsigned; the server logs that mode loudly at startup."""
    if not secret:
        return Verification(True, "unsigned", "no secret configured for this provider")
    if provider == "github":
        return _verify_hmac(header(headers, GITHUB_SIGNATURE_HEADER), body, secret, GITHUB_SIGNATURE_HEADER)
    if provider == "gitcode":
        sig = header(headers, GITCODE_SIGNATURE_HEADER)
        if sig:
            return _verify_hmac(sig, body, secret, GITCODE_SIGNATURE_HEADER, allow_base64=True)
        token = header(headers, GITCODE_TOKEN_HEADER)
        if token is not None:
            if hmac.compare_digest(token.encode("utf-8"), secret.encode("utf-8")):
                return Verification(True, "token")
            return Verification(False, "rejected", f"{GITCODE_TOKEN_HEADER} does not match")
        return Verification(False, "rejected", f"missing {GITCODE_SIGNATURE_HEADER} or {GITCODE_TOKEN_HEADER}")
    return Verification(False, "rejected", f"unknown provider {provider!r}")


def _verify_hmac(value: str | None, body: bytes, secret: str, name: str, allow_base64: bool = False) -> Verification:
    if not value:
        return Verification(False, "rejected", f"missing {name}")
    scheme, _, digest = value.strip().partition("=")
    if scheme.lower() != "sha256" or not digest:
        return Verification(False, "rejected", f"malformed {name}: expected sha256=<digest>")
    mac = hmac.new(secret.encode("utf-8"), body, hashlib.sha256)
    candidates = [mac.hexdigest()]
    if allow_base64:
        candidates.append(base64.b64encode(mac.digest()).decode("ascii"))
    if any(hmac.compare_digest(candidate, digest) for candidate in candidates):
        return Verification(True, "hmac-sha256")
    return Verification(False, "rejected", f"{name} mismatch")


def sign(body: bytes, secret: str) -> str:
    """Header value GitHub would send for ``body``; also what the GitCode adapter accepts."""
    return "sha256=" + hmac.new(secret.encode("utf-8"), body, hashlib.sha256).hexdigest()

File: src/core/signature.ts
import { utf8, type Verification } from './types.js';

export const detectProvider = (headers: Headers): string => headers.has('x-github-event') ? 'github' : headers.has('x-gitcode-event') ? 'gitcode' : '';
export const hex = (bytes: Uint8Array): string => Array.from(bytes, n => n.toString(16).padStart(2, '0')).join('');
export function base64(bytes: Uint8Array): string {
  let result = '';
  for (let i = 0; i < bytes.length; i += 8192) result += String.fromCharCode(...bytes.subarray(i, i + 8192));
  return btoa(result);
}
export const fromBase64 = (value: string): Uint8Array<ArrayBuffer> => Uint8Array.from(atob(value), c => c.charCodeAt(0));
export async function digest(bytes: Uint8Array): Promise<string> {
  return hex(new Uint8Array(await crypto.subtle.digest('SHA-256', new Uint8Array(bytes))));
}
async function hmacKey(secret: string): Promise<CryptoKey> {
  return crypto.subtle.importKey('raw', utf8.encode(secret), { name: 'HMAC', hash: 'SHA-256' }, false, ['sign', 'verify']);
}
export async function sign(body: Uint8Array, secret: string): Promise<string> {
  return 'sha256=' + hex(new Uint8Array(await crypto.subtle.sign('HMAC', await hmacKey(secret), new Uint8Array(body))));
}
export async function equalSecret(value: string, expected: string): Promise<boolean> {
  // HMAC verification avoids timing comparisons of variable-length bearer/token strings.
  const key = await hmacKey(expected);
  const signature = await crypto.subtle.sign('HMAC', key, utf8.encode(expected));
  return crypto.subtle.verify('HMAC', key, signature, utf8.encode(value));
}
export async function verify(provider: string, headers: Headers, body: Uint8Array, secret?: string): Promise<Verification> {
  if (!secret) return { ok: true, mode: 'unsigned', reason: 'no secret configured' };
  if (!['github', 'gitcode'].includes(provider)) return { ok: false, mode: 'rejected', reason: 'unknown provider' };
  const signature = headers.get(provider === 'github' ? 'x-hub-signature-256' : 'x-gitcode-signature-256');
  if (signature) {
    let bytes: Uint8Array<ArrayBuffer> | undefined;
    const encoded = /^sha256=(.+)$/i.exec(signature)?.[1] || '';
    if (/^[0-9a-f]{64}$/.test(encoded)) bytes = Uint8Array.from(encoded.match(/../g)!, v => parseInt(v, 16));
    else if (provider === 'gitcode' && /^[A-Za-z0-9+/]{43}=$/.test(encoded)) {
      const decoded = fromBase64(encoded); if (base64(decoded) === encoded) bytes = decoded;
    }
    const ok = !!bytes && await crypto.subtle.verify('HMAC', await hmacKey(secret), bytes, new Uint8Array(body));
    return { ok, mode: ok ? 'hmac-sha256' : 'rejected', reason: ok ? '' : 'signature mismatch' };
  }
  if (provider === 'gitcode' && headers.has('x-gitcode-token')) {
    const ok = await equalSecret(headers.get('x-gitcode-token')!, secret);
    return { ok, mode: ok ? 'token' : 'rejected', reason: ok ? '' : 'token mismatch' };
  }
  return { ok: false, mode: 'rejected', reason: 'missing signature' };
}
